File: invictus_android/android/src/main/kotlin/dev/zeekr/invictus_android/net/util/Inet4AddressUtils.kt
package dev.zeekr.invictus_android.net.util

import java.net.Inet4Address
import java.net.InetAddress
import java.net.UnknownHostException

/**
 * Collection of utilities to work with IPv4 addresses.
 */
object Inet4AddressUtils {
    internal val clazz: Class<*> = Class.forName("com.android.net.module.util.Inet4AddressUtils")

    /**
     * Convert a IPv4 address from an integer to an InetAddress (0x04030201 -> 1.2.3.4)
     *
     *
     * This method uses the higher-order int bytes as the lower-order IPv4 address bytes,
     * which is an unusual convention. Consider [.intToInet4AddressHTH] instead.
     * @param hostAddress an int coding for an IPv4 address, where higher-order int byte is
     * lower-order IPv4 address byte
     */
    fun intToInet4AddressHTL(hostAddress: Int): Inet4Address {
        return clazz.getMethod("intToInet4AddressHTL", Int::class.java)
            .invoke(null, hostAddress) as Inet4Address
    }

    /**
     * Convert a IPv4 address from an integer to an InetAddress (0x01020304 -> 1.2.3.4)
     * @param hostAddress an int coding for an IPv4 address
     */
    fun intToInet4AddressHTH(hostAddress: Int): Inet4Address {
        return clazz.getMethod("intToInet4AddressHTH", Int::class.java)
            .invoke(null, hostAddress) as Inet4Address
    }

    /**
     * Convert an IPv4 address from an InetAddress to an integer (1.2.3.4 -> 0x01020304)
     *
     *
     * This conversion can help order IP addresses: considering the ordering
     * 192.0.2.1 < 192.0.2.2 < ..., resulting ints will follow that ordering if read as unsigned
     * integers with [Integer.toUnsignedLong].
     * @param inetAddr is an InetAddress corresponding to the IPv4 address
     * @return the IP address as integer
     */
    fun inet4AddressToIntHTH(inetAddr: Inet4Address): Int {
        return clazz.getMethod("inet4AddressToIntHTH", Inet4Address::class.java)
            .invoke(null, inetAddr) as Int
    }

    /**
     * Convert a IPv4 address from an InetAddress to an integer (1.2.3.4 -> 0x04030201)
     *
     *
     * This method stores the higher-order IPv4 address bytes in the lower-order int bytes,
     * which is an unusual convention. Consider [.inet4AddressToIntHTH] instead.
     * @param inetAddr is an InetAddress corresponding to the IPv4 address
     * @return the IP address as integer
     */
    fun inet4AddressToIntHTL(inetAddr: Inet4Address): Int {
        return clazz.getMethod("inet4AddressToIntHTL", Inet4Address::class.java)
            .invoke(null, inetAddr) as Int
    }

    /**
     * Convert a network prefix length to an IPv4 netmask integer (prefixLength 17 -> 0xffff8000)
     * @return the IPv4 netmask as an integer
     */
    fun prefixLengthToV4NetmaskIntHTH(prefixLength: Int): Int {
        return clazz.getMethod("prefixLengthToV4NetmaskIntHTH", Int::class.java)
            .invoke(null, prefixLength) as Int
    }

    /**
     * Convert a network prefix length to an IPv4 netmask integer (prefixLength 17 -> 0x0080ffff).
     *
     *
     * This method stores the higher-order IPv4 address bytes in the lower-order int bytes,
     * which is an unusual convention. Consider [.prefixLengthToV4NetmaskIntHTH] instead.
     * @return the IPv4 netmask as an integer
     */
    fun prefixLengthToV4NetmaskIntHTL(prefixLength: Int): Int {
        return clazz.getMethod("prefixLengthToV4NetmaskIntHTL", Int::class.java)
            .invoke(null, prefixLength) as Int
    }

    /**
     * Convert an IPv4 netmask to a prefix length, checking that the netmask is contiguous.
     * @param netmask as a `Inet4Address`.
     * @return the network prefix length
     * @throws IllegalArgumentException the specified netmask was not contiguous.
     */
    fun netmaskToPrefixLength(netmask: Inet4Address): Int {
        return clazz.getMethod("netmaskToPrefixLength", Inet4Address::class.java)
            .invoke(null, netmask) as Int
    }

    /**
     * Returns the implicit netmask of an IPv4 address, as was the custom before 1993.
     */
    fun getImplicitNetmask(address: Inet4Address): Int {
        return clazz.getMethod("getImplicitNetmask", Inet4Address::class.java)
            .invoke(null, address) as Int
    }

    /**
     * Get the broadcast address for a given prefix.
     *
     *
     * For example 192.168.0.1/24 -> 192.168.0.255
     */
    fun getBroadcastAddress(addr: Inet4Address, prefixLength: Int): Inet4Address {
        return clazz.getMethod("getBroadcastAddress", Inet4Address::class.java, Int::class.java)
            .invoke(null, addr, prefixLength) as Inet4Address
    }

    /**
     * Get a prefix mask as Inet4Address for a given prefix length.
     *
     *
     * For example 20 -> 255.255.240.0
     */
    fun getPrefixMaskAsInet4Address(prefixLength: Int): Inet4Address {
        return clazz.getMethod("getPrefixMaskAsInet4Address", Int::class.java)
            .invoke(null, prefixLength) as Inet4Address
    }

    /**
     * Trim leading zeros from IPv4 address strings
     * Non-v4 addresses and host names remain unchanged.
     * For example, 192.168.000.010 -> 192.168.0.10
     * @param addr a string representing an ip address
     * @return a string properly trimmed
     */
    fun trimAddressZeros(addr: String?): String? {
        return clazz.getMethod("trimAddressZeros", String::class.java).invoke(null, addr) as String?
    }
}

File: invictus_api/lib/src/net/util/inet4_address_utils.dart
import 'package:invictus_api/src/net.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

abstract interface class Inet4AddressUtils {
  static Inet4Address intToInet4AddressHTL(int hostAddress) =>
      Inet4AddressUtilsChannel.instance.intToInet4AddressHTL(hostAddress);

  static Inet4Address intToInet4AddressHTH(int hostAddress) =>
      Inet4AddressUtilsChannel.instance.intToInet4AddressHTH(hostAddress);

  static int inet4AddressToIntHTH(Inet4Address inetAddr) =>
      Inet4AddressUtilsChannel.instance.inet4AddressToIntHTH(inetAddr);

  static int inet4AddressToIntHTL(Inet4Address inetAddr) =>
      Inet4AddressUtilsChannel.instance.inet4AddressToIntHTL(inetAddr);

  static int prefixLengthToV4NetmaskIntHTH(int prefixLength) =>
      Inet4AddressUtilsChannel.instance.prefixLengthToV4NetmaskIntHTH(
        prefixLength,
      );

  static int prefixLengthToV4NetmaskIntHTL(int prefixLength) =>
      Inet4AddressUtilsChannel.instance.prefixLengthToV4NetmaskIntHTL(
        prefixLength,
      );

  static int netmaskToPrefixLength(Inet4Address netmask) =>
      Inet4AddressUtilsChannel.instance.netmaskToPrefixLength(netmask);

  static int getImplicitNetmask(Inet4Address address) =>
      Inet4AddressUtilsChannel.instance.getImplicitNetmask(address);

  static Inet4Address getBroadcastAddress(
    Inet4Address addr,
    int prefixLength,
  ) => Inet4AddressUtilsChannel.instance.getBroadcastAddress(addr, prefixLength);

  static Inet4Address getPrefixMaskAsInet4Address(int prefixLength) =>
      Inet4AddressUtilsChannel.instance.getPrefixMaskAsInet4Address(
        prefixLength,
      );

  static String? trimAddressZeros(String? addr) =>
      Inet4AddressUtilsChannel.instance.trimAddressZeros(addr);
}

abstract base class Inet4AddressUtilsChannel extends PlatformInterface {
  /// Constructs a [Inet4AddressUtilsChannel].
  Inet4AddressUtilsChannel() : super(token: _token);

  static final Object _token = Object();

  static Inet4AddressUtilsChannel? _instance;

  /// The default instance of [Inet4AddressUtilsChannel] to use.
  static Inet4AddressUtilsChannel get instance =>
      ArgumentError.checkNotNull(_instance, 'instance');

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [Inet4AddressUtilsChannel] when
  /// they register themselves.
  static set instance(Inet4AddressUtilsChannel instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  Inet4Address intToInet4AddressHTL(int hostAddress);
  Inet4Address intToInet4AddressHTH(int hostAddress);
  int inet4AddressToIntHTH(Inet4Address inetAddr);
  int inet4AddressToIntHTL(Inet4Address inetAddr);
  int prefixLengthToV4NetmaskIntHTH(int prefixLength);
  int prefixLengthToV4NetmaskIntHTL(int prefixLength);
  int netmaskToPrefixLength(Inet4Address netmask);
  int getImplicitNetmask(Inet4Address address);
  Inet4Address getBroadcastAddress(Inet4Address addr, int prefixLength);
  Inet4Address getPrefixMaskAsInet4Address(int prefixLength);
  String? trimAddressZeros(String? addr);
}
